File: lib/utils/sigatoka_date_util.dart
/// Utilidad para calcular semana epidemiológica ISO y período del mes
class SigatokaDateUtil {
  /// Obtiene la semana epidemiológica ISO 8601
  /// La semana 1 es aquella que contiene el primer jueves del año
  /// Las semanas empiezan en lunes
  static int getSemanaEpidemiologicaISO(DateTime fecha) {
    // Encontrar el lunes de la semana actual
    // En Dart: Monday = 1, Sunday = 7
    int diasDesdeElLunes = fecha.weekday - 1;
    DateTime lunesDeLaSemana = fecha.subtract(Duration(days: diasDesdeElLunes));
    
    // Encontrar el primer jueves del año
    // La semana 1 es la que contiene el primer jueves
    DateTime primerEnero = DateTime(lunesDeLaSemana.year, 1, 1);
    int diasHastaJueves = (DateTime.thursday - primerEnero.weekday + 7) % 7;
    DateTime primerJueves = primerEnero.add(Duration(days: diasHastaJueves));
    
    // Encontrar el lunes de la semana del primer jueves
    DateTime lunesDeSemana1 = primerJueves.subtract(Duration(days: primerJueves.weekday - 1));
    
    // Si la fecha es antes del lunes de la semana 1, pertenece al año anterior
    if (lunesDeLaSemana.isBefore(lunesDeSemana1)) {
      // Calcular semana del año anterior
      return getSemanaEpidemiologicaISO(DateTime(lunesDeLaSemana.year - 1, 12, 28));
    }
    
    // Calcular número de semana
    int diasDesdeSemana1 = lunesDeLaSemana.difference(lunesDeSemana1).inDays;
    int numeroSemana = (diasDesdeSemana1 / 7).floor() + 1;
    
    // Si es semana 53, verificar si realmente existe o es semana 1 del siguiente año
    if (numeroSemana == 53) {
      DateTime primerEneroSiguiente = DateTime(lunesDeLaSemana.year + 1, 1, 1);
      int diasHastaJuevesSiguiente = (DateTime.thursday - primerEneroSiguiente.weekday + 7) % 7;
      DateTime primerJuevesSiguiente = primerEneroSiguiente.add(Duration(days: diasHastaJuevesSiguiente));
      DateTime lunesDeSemana1Siguiente = primerJuevesSiguiente.subtract(Duration(days: primerJuevesSiguiente.weekday - 1));
      
      if (lunesDeLaSemana.isAtSameMomentAs(lunesDeSemana1Siguiente) || 
          lunesDeLaSemana.isAfter(lunesDeSemana1Siguiente)) {
        return 1;
      }
    }
    
    return numeroSemana;
  }
  
  /// Calcula el período como "Semana X de Mes Y"
  static String getPeriodoSemanaDelMes(DateTime fecha) {
    final int diaDelMes = fecha.day;
    final int semanaDelMes = ((diaDelMes - 1) ~/ 7) + 1;
    final String mes = getMesEnEspanol(fecha.month);
    
    return 'Semana $semanaDelMes de $mes';
  }
  
  /// Obtiene la semana del mes (1-5)
  static int getSemanaDelMes(DateTime fecha) {
    final int diaDelMes = fecha.day;
    return ((diaDelMes - 1) ~/ 7) + 1;
  }
  
  /// Obtiene el nombre del mes en español
  static String getMesEnEspanol(int numeroMes) {
    const List<String> meses = [
      'Enero', 'Febrero', 'Marzo', 'Abril', 'Mayo', 'Junio',
      'Julio', 'Agosto', 'Septiembre', 'Octubre', 'Noviembre', 'Diciembre'
    ];
    return meses[numeroMes - 1];
  }
  
  /// Formatea una fecha completa con semana ISO y período
  static String getFormatoCompleto(DateTime fecha) {
    final int semanaISO = getSemanaEpidemiologicaISO(fecha);
    final String periodo = getPeriodoSemanaDelMes(fecha);
    return 'Semana ISO: $semanaISO, Período: $periodo';
  }
}

File: backend_new/src/main/java/com/lytiks/backend/util/SigatokaDateUtil.java
package com.lytiks.backend.util;

import java.time.LocalDate;
import java.time.temporal.IsoFields;
import java.time.temporal.WeekFields;
import java.util.Locale;

/**
 * Utilidades para cálculos de fechas en Sigatoka
 */
public class SigatokaDateUtil {
    
    /**
     * Obtiene la semana epidemiológica ISO 8601
     * @param fecha Fecha para calcular la semana
     * @return Número de semana ISO (1-53)
     */
    public static int getSemanaEpidemiologicaISO(LocalDate fecha) {
        return fecha.get(IsoFields.WEEK_OF_WEEK_BASED_YEAR);
    }
    
    /**
     * Calcula el período como "Semana X del Mes Y"
     * @param fecha Fecha para calcular el período
     * @return String con el formato "Semana 1 del Mes", "Semana 2 del Mes", etc.
     */
    public static String getPeriodoSemanaDelMes(LocalDate fecha) {
        // Obtener el día del mes
        int diaDelMes = fecha.getDayOfMonth();
        
        // Calcular la semana del mes (1-5)
        int semanaDelMes = ((diaDelMes - 1) / 7) + 1;
        
        // Obtener el nombre del mes en español
        String mes = getMesEnEspanol(fecha.getMonthValue());
        
        return String.format("Semana %d de %s", semanaDelMes, mes);
    }
    
    /**
     * Obtiene la semana del mes (1-5)
     * @param fecha Fecha para calcular
     * @return Número de semana del mes
     */
    public static int getSemanaDelMes(LocalDate fecha) {
        int diaDelMes = fecha.getDayOfMonth();
        return ((diaDelMes - 1) / 7) + 1;
    }
    
    /**
     * Obtiene el nombre del mes en español
     * @param numeroMes Número del mes (1-12)
     * @return Nombre del mes en español
     */
    public static String getMesEnEspanol(int numeroMes) {
        String[] meses = {
            "Enero", "Febrero", "Marzo", "Abril", "Mayo", "Junio",
            "Julio", "Agosto", "Septiembre", "Octubre", "Noviembre", "Diciembre"
        };
        return meses[numeroMes - 1];
    }
    
    /**
     * Obtiene el año de la semana ISO
     * @param fecha Fecha para calcular
     * @return Año de la semana ISO
     */
    public static int getAnioSemanaISO(LocalDate fecha) {
        return fecha.get(IsoFields.WEEK_BASED_YEAR);
    }
    
    /**
     * Formatea una fecha completa con semana ISO y período
     * @param fecha Fecha para formatear
     * @return String con formato: "Semana ISO: X, Período: Semana Y de Mes"
     */
    public static String getFormatoCompleto(LocalDate fecha) {
        int semanaISO = getSemanaEpidemiologicaISO(fecha);
        String periodo = getPeriodoSemanaDelMes(fecha);
        return String.format("Semana ISO: %d, Período: %s", semanaISO, periodo);
    }
}
